proposals/0000-wire-protocol.md

```diff
@@ -0,0 +1,435 @@
+
+By default, a peer wants no blocks. It can add or remove wanted block using the Want and Unwant messages.
+
+When new wanted blocks are made available, the peer should react by sending a Have message. Likewise, when wanted blocks are no longer available, the peer should react by sending an Unhave message. Finally, any time a Want message is received, the peer should react by sending a Have message.
+
+
+## Request Procedure
+[requests]: #requests
+
+After the Want/Have Procedure, a peer will know what Hypercore feed blocks are available on the remote, and can then send Request messages for the individual blocks. Each Request specifies which block it needs, and also specifies some information about Merkle tree proof nodes which should be included. In reaction, the remote sends a Data message with the requested content.
+
+At present, Request messages can only specify one block at a time. This is to encourage an equal distribution of requests between multiple connected peers. However, multiple requests can be sent in parallel.
+
+
+## Extension Procedure
+[extensions]: #extensions
+
+To allow for experimentation within userland, the wire protocol supports an extension process. All extensions are identified by strings, and sent in an array in the Handshake message. Both peers must declare an extension in the handshake to consider it 'supported'.
+
+Each extension is capable of sending custom payloads through the Extension message type.
+
+
+# Message Details
+[message-details]: #message-details
+
+TODO: description of framing
+
+Wire format is `<len>(<header><message>)`. `header` is a varint, of form `channel << 4 | <4-bit-type>`.
+
+Messages are encoded (serialized) using Google's [protobuf][protobuf] encoding.
+
+[protobuf]: https://developers.google.com/protocol-buffers/
+
+<table>
+  <tr><th>`type` code <th>Name
+  <tr><td> N/A <td>[Keep-Alive][msg-keepalive]
+  <tr><td> 0 <td>[Feed][msg-feed]
+  <tr><td> 1 <td>[Handshake][msg-handshake]
+  <tr><td> 2 <td>[Info][msg-info]
+  <tr><td> 3 <td>[Have][msg-have]
+  <tr><td> 4 <td>[Unhave][msg-unhave]
+  <tr><td> 5 <td>[Want][msg-want]
+  <tr><td> 6 <td>[Unwant][msg-unwant]
+  <tr><td> 7 <td>[Request][msg-request]
+  <tr><td> 8 <td>[Cancel][msg-cancel]
+  <tr><td> 9 <td>[Data][msg-data]
+  <tr><td>15 <td>[Extension][msg-extension]
+</table>
+
+#### Keep-Alive
+[msg-keepalive]: #msg-keepalive
+
+A message of body length 0 (giving a total message size of 1 byte for the `len` varint) is a keep-alive. Depending on transport and application needs, peers may optionally send keep-alive messages to help detect and prevent connection loss. Peers must always handle keep-alive messages correctly (aka, ignore them), regardless of transport.
+
+TODO: what is a good default interval?
```

Comment:
From brief internet reading, sounds like surprisingly long (300 seconds / 5 minutes) works to resolve the operating system socket use-case.
I don't think it's crazy to imagine a server (like hashbase) maintaining 1000 active connections; with a 300 second interval that's already several per second.